core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -619,6 +625,47 @@ private[spark] class ExecutorAllocationManager(
     // place the executors.
     private val stageIdToExecutorPlacementHints = new mutable.HashMap[Int, (Int, Map[String, Int])]
 
+    override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+      jobStart.stageInfos.foreach(stageInfo => stageIdToJobId(stageInfo.stageId) = jobStart.jobId)
+
+      var jobGroupId = if (jobStart.properties != null) {
+        jobStart.properties.getProperty(SparkContext.SPARK_JOB_GROUP_ID)
+      } else {
+        null
+      }
+
+      val maxConTasks = if (jobGroupId != null &&
+        conf.contains(s"spark.job.$jobGroupId.maxConcurrentTasks")) {
+        conf.get(s"spark.job.$jobGroupId.maxConcurrentTasks").toInt
+      } else {
+        Int.MaxValue
+      }
+
+      if (maxConTasks <= 0) {
+        throw new IllegalArgumentException(
+          "Maximum Concurrent Tasks should be set greater than 0 for the job to progress.")
+      }
+
+      if (jobGroupId == null || !conf.contains(s"spark.job.$jobGroupId.maxConcurrentTasks")) {
+        jobGroupId = DEFAULT_JOB_GROUP
+      }
+
+      jobIdToJobGroup(jobStart.jobId) = jobGroupId
+      if (!jobGroupToMaxConTasks.contains(jobGroupId)) {
```

Comment:
this is probably a weird / unusual situation, but is this really the behavior you want if there are multiple jobs submitted for the same job group?  Wouldn't you just take the conf for the job group at the time each job was submitted?

Worst case with this approach: say you are *always* submitting multiple jobs for each job group; when one finishes, you immediately start another one, so that the new one partially overlaps the old one.  Then even if you change the conf, all jobs will keep using the old value forever.

Reply:
From my understanding and in context of the current code, you would group jobs together when you want their concurrency to be limited. If you want different concurrency limits for different jobs, you would set them in a different jobgroup altogether. 

If there are multiple jobs in the same job group which run concurrently and one of them sets a value different, then which one wins for the existing jobs and the new job? If we want to have a different value for every job then the user would need a way to know and identify a spark job in his application code , probably by a job id. Only by means of identifying a job, would the user be able to set the config for that job. This cannot be known apriori and I don't know if there is an easy way that the user can know about the underlying spark job corresponding to the action. Hence we apply a setting at the jobgroup level which allows the user to allow him to control the concurrency without knowing the underlying job related details specific to spark in an easy manner.

Let me know if anything is unclear here or if you have more questions.

Reply:
lemme put my concern another way:

why don't you remove the `if (!jobGroupToMaxConTasks.contains(jobGroupId))`, and just unconditionally always make the assignment `jobGroupToMaxConTasks(jobGroupId) = maxConTasks`?

that is simpler to reason about, and has all the properties we want.  I agree the scenario I'm describing is pretty weird, but the only difference I see between your version and this is in that scenario.  And its probably not the behavior we want in that scenario.

Reply:
I understand your reasoning about setting the maxConTasks every time the job is set. However, I am not able to understand the scenario which you described. If a job completes and new one kicks off immediately, how does the new job partially overlap? Its only when all the stages & underlying tasks for the previous job have finished we would mark it as complete. So a new job won't overlap with a completed one. Am I missing something here?

Reply:
You could submit jobs concurrently, from two different threads.  I didn't describe it very well -- say both threads are in the same job group.  Each thread can only do one job at a time, but maybe between the two threads, there is always some active job for the job group the entire time.

There are real use cases which are like this -- in "job-server" style deployments, there is a long-running spark context (most likely with cached RDDs), accepting requests from multiple users (perhaps sitting behind an http server).  Maybe some group of users are always put into one job group (eg., there is a low-priority group and a high priority group of users).  You might process more than one job for each group at a time, and there is a never ending stream of jobs.

(again, its not the most common scenario, but might as well have it behave correctly.)

Reply:
Okay. We did consider the job-server style deployments, however I am not able to follow the example/scenario you mentioned.

So when you say that different set of users are assigned different groups, they have to be served by "different threads", because we can set the job group on a per thread basis only. 

While it is possible & valid for different threads to belong to the same job group, you cannot have multiple job groups for the same thread active simultaneously. 

So in this scenario where you would want to change the maxConcTasks for a given jobGroup in one thread, the changes would be visible in the other thread as well. Because both of them belong to the same job group. In this case, it seems that it wasn't configured correctly. I am still not able to follow how setting the value every single time will help here.

Reply:
If we are talking jobs within the same job groups, it seems like this would be very timing dependent as to what number you would get if you start allowing it to be changed real time.  Lets say you have 1 thread and set the job group.  Now if all the jobs within that group are launched serial then everything is easy, allowing it to be changed can make sense. But if from that thread you spawn other threads to launch jobs in parallel (which would still be in that same job group) and each of those is setting it differently, how do you know you will get the right number for each of those jobs?   the 2 threads could race to set the conf and if both set it right before launching you are going to get one of the settings for both launches whereas one might have expected a different setting.

@squito  does this cover the scenario you are referring to?  

while both of those cases might be rare, I would lean towards making sure its more predictable and only setting it once rather then having user get something they don't expect.  But either could probably be documented away if we see the serial type scenario being more beneficial.  

ideally it would be nice to set at the stage level but that is a lot more difficult.

Reply:
Sorry I am doing a really bad job explaining my concerns.

Of course, if you have multiple threads racing to set the value for the same job group, its unpredictable which value you'll get.  Nothing we can do about that, its not the scenario I'm talking about.

Here's the step-by-step (hope I understand the proposed change correctly):

(1) T1: Set jobgroup = "foo".  Set maxConTasks=10.
(2) T1: Launch job 1.1  Uses maxConTasks=10
(3) T2: Set jobgroup = "foo".
(4) T2: Launch job 2.1. Uses maxConTasks=10
(5) T1:  Finish job 1.1.  Do *not* remove the entry for `jobGroupToMaxConTasks("foo")`, because there is still another job running for this job group in T2 (https://github.com/apache/spark/pull/19194/files#diff-b096353602813e47074ace09a3890d56R664)
(6) T1: Set maxConTasks=20
(7) T1: Launch job 1.2.  Uses maxConTasks=10, because `jobGroupToMaxConTasks.contains("foo")`, so we don't reset the value.
(8) T2: Finish job 2.1.  Again, do not remove `jobGroupToMaxConTasks("foo")`, as T1 is still running a job in this job group.
(9) Set maxConTasks=20
(10) T2: Run job 2.2.  Uses maxConTasks=10, because `jobGroupToMaxConTasks.contains("foo")`

As long as there a job running in T2 when T1 finishes its job (or vice versa), we never remove the prior value, and so never update it when starting a new job.  We could be stuck with maxConTasks=10 in both threads indefinitely, even though both threads have set maxConTasks=20.

If you remove that `if (!jobGroupToMaxConTasks.contains(jobGroupId))`, then when job 1.2 starts, you'd use the new value of maxConTasks=20.  The only weird thing is that job 2.1 suddenly switches mid-flight to using maxConTasks=20 as well.  But that seems more reasonable to me.

Reply:
Do we even need mutable state associated with a job group? Some things would be a lot simpler if maxConTasks could only be set when the job group is created; and if you need a different number of maxConTasks, then you have to use a different job group.

Reply:
@squito Thanks for elaborating on this. I believe the last para of your explanation summarized your viewpoint. In a way that makes some sense because if you want to increase your max concurrent tasks, you know what you are doing, so if you see weird thing with other threads that you created, its fine. However, when you don't control the thread creation, I feel its best to set it just once to avoid the weirdness. Its much easier to use a different job group than explain one more weird behavior.